psydac/linalg/kron.py

```diff
@@ -217,11 +232,170 @@ def transpose(self, conjugate=False):
+
+    # ...
+    @property
+    def dtype(self):
+        return self.domain.dtype
+
+    # ...
+    @property
+    def ndim(self):
+        return self._ndim
+
+    # ...
     @property
-    def T(self):
-        return self.transpose()
+    def mats(self):
+        return self._mats
+
+    # ...
+    def dot(self, x, out=None):
+
+        dot = np.dot
+
+        assert isinstance(x, StencilVector)
+        assert x.space is self.domain
 
-class KroneckerLinearSolver( LinearSolver ):
+        # Necessary if vector space is periodic or distributed across processes
+        if not x.ghost_regions_in_sync:
+            x.update_ghost_regions()
+
+        if out is not None:
+            assert isinstance(out, StencilVector)
+            assert out.space is self.codomain
+        else:
+            out = StencilVector(self.codomain)
+
+        d_starts = self._domain.starts
+        d_ends   = self._domain.ends
+        c_starts = self._codomain.starts
+        c_ends   = self._codomain.ends
+        pads     = self._codomain.pads
+        mats     = self.mats
+
+        nrows  = tuple(e-s+1 for s,e in zip(c_starts, c_ends))
+        ncols  = tuple(e-s+1+2*p for s,e,p in zip(d_starts, d_ends, pads))
+        kk     = tuple(slice(s, s+nc) for nc,s in zip(ncols, d_starts))
+        x_data   = x._data.ravel()
+        out_data = out._data
+
+        for xx in np.ndindex(*nrows):
+            ii     = tuple(x+p for x,p in zip(xx,pads))
+            i_mats = [mat[i+s, k] for i,s,k,mat in zip(ii, c_starts, kk, mats)]
+            out_data[ii] = np.dot(x_data, np.outer(*i_mats).ravel())
+
+        # IMPORTANT: flag that ghost regions are not up-to-date
+        out.ghost_regions_in_sync = False
+        return out
+
+    # ...
+    def copy(self):
+        mats = [m.copy() for m in self.mats]
+        return KroneckerDenseMatrix(self.domain, self.codomain, *mats, with_pads=True)
+
+    # ...
+    def __neg__(self):
+        mats = [-self.mats[0], *(m.copy() for m in self.mats[1:])]
+        return KroneckerDenseMatrix(self.domain, self.codomain, *mats, with_pads=True)
+
+    # ...
+    def __mul__(self, a):
+        mats = [*(m.copy() for m in self.mats[:-1]), self.mats[-1] * a]
+        return KroneckerDenseMatrix(self.domain, self.codomain, *mats, with_pads=True)
+
+    # ...
+    def __rmul__(self, a):
+        mats = [a * self.mats[0], *(m.copy() for m in self.mats[1:])]
+        return KroneckerDenseMatrix(self.domain, self.codomain, *mats, with_pads=True)
+
+    # ...
+    def __imul__(self, a):
+        self.mats[-1] *= a
+        return self
+
+    # ...
+    def __add__(self, m):
+        raise NotImplementedError('Cannot sum Kronecker matrices')
+
+    def __sub__(self, m):
+        raise NotImplementedError('Cannot subtract Kronecker matrices')
+
+    def __iadd__(self, m):
+        raise NotImplementedError('Cannot sum Kronecker matrices')
+
+    def __isub__(self, m):
+        raise NotImplementedError('Cannot subtract Kronecker matrices')
```

Comment:
I see no reason to raise these errors, because the `LinearOperator` base class already provides the methods `__add__` and `__sub__`. The same comment applies to class `KroneckerStencilMatrix` in the same file.

I am not 100% of what we should do with `__iadd__` and `__isub__` though. @jowezarek Any suggestions?

Reply:
done in [d9aabe8](https://github.com/pyccel/psydac/pull/241/commits/d9aabe882eb9631a8e9256270bef4ec3d0b63889)